### PR TITLE
Create EntityNotDeclared code action

### LINKED ISSUE #532: CodeAction for EntityNotDeclared
(Originally from #518 )

Given this XML file:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE article [
  <!ELEMENT article (#PCDATA)>
]>          
<article>
	&nbsp;
</article>
```

&nbsp; should be highlighted. (It should be managed in XMLSyntaxErrorCode).

CodeAction could create the proper entity in the DOCTYPE like this:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE article [
  <!ELEMENT article (#PCDATA)>
  <!ENTITY nbsp "entity-value">
]>           
<article>
	&nbsp;
</article>
```
We must take care of existing of doctype. For instance if we have

```xml
<?xml version="1.0" encoding="UTF-8"?>          
<article>
	&nbsp;
</article>
```

CodeAction should generate

```xml
<!DOCTYPE article [
  <!ENTITY nbsp "entity-value">
]>
```

### PR DESCRIPTION
**Pull request by xorye**
Fixes #532 

This PR adds a new code action to locally declare an undeclared entity.

```xml
<?xml version="1.0" encoding="UTF-8"?>
<article>
   &nbsp;
</article>
```

Demo:
demo

The codeaction works whether or not `<!DOCTYPE` is declared or not, and whether the internal subset `[]` exists or not.

Signed-off-by: David Kwon <dakwon@redhat.com>

### CODE REVIEW

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/EntityNotDeclaredCodeAction.java [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+		insertString.append(delimiter);
+		insertString.append("]>" + delimiter);
+		CodeAction action = CodeActionFactory.insert(message, insertPosition, insertString.toString(),
+				document.getTextDocument(), diagnostic);
+		codeActions.add(action);
+	}
+
+	private String getEntityName(Diagnostic diagnostic, Range range, DOMDocument doc) throws BadLocationException {
+		String name = doc.getText().substring(doc.offsetAt(range.getStart()), doc.offsetAt(range.getEnd()));
+		String removedAmpAndSemiColon = name.substring(1, name.length() - 1);
```
Comment (xorye) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/EntityNotDeclaredCodeAction.java:157]: Here, I am retrieving the entity name from the error range, and then checking it against the error message.
It would be better if we could store the name in the `Diagnostic` when the `Diagnostic` is being created.

Related issue: https://github.com/eclipse/lemminx/issues/681

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/EntityNotDeclaredCodeAction.java [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+			insertPosition = new Position(0, 0);
+		}
+
+		String delimiter = document.lineDelimiter(insertPosition.getLine());
+		String message = "Declare DOCTYPE containing ENTITY " + entityName;
+		StringBuilder insertString = new StringBuilder();
+		insertString.append("<!DOCTYPE article [" + delimiter);
+		insertString.append(createEntityDeclaration(entityName, settings));
+		insertString.append(delimiter);
+		insertString.append("]>" + delimiter);
```
Comment (xorye) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/EntityNotDeclaredCodeAction.java:149]: TBH is it okay to use string concatenation instead of StringBuilder in this PR? I think StringBuilder might not be necessary unless I am appending string content in a loop.

https://stackoverflow.com/questions/4645020/when-to-use-stringbuilder-in-java
https://stackoverflow.com/questions/5234147/why-stringbuilder-when-there-is-string
https://stackoverflow.com/questions/24571463/when-should-you-explicitly-use-a-stringbuilder

File: org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/extensions/contentmodel/DTDDiagnosticsTest.java [hunk 3]
```diff
@@ -280,6 +285,143 @@ public void EntityDeclUnterminated() throws Exception {
...
+				"	&nbsp;\r\n" + //
+				"</article>";
+		SharedSettings settings = new SharedSettings();
+		settings.getPreferences().setQuoteStyle(QuoteStyle.singleQuotes);
+		settings.getFormattingSettings().setInsertSpaces(false);
+		Diagnostic d = d(5, 1, 5, 7, DTDErrorCode.EntityNotDeclared, "The entity \"nbsp\" was referenced, but not declared.");
+		XMLAssert.testDiagnosticsFor(xml, d);
+		testCodeActionsFor(xml, d, settings, ca(d, te(2, 29, 2, 29, "\r\n\t<!ENTITY nbsp \'entity-value\'>")));
+	}
+
```
Comment (datho7561) [inline org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/extensions/contentmodel/DTDDiagnosticsTest.java:424]:
```xml
<?xml version="1.0" encoding="UTF-8"?><text>
<text>
  &c;
</text>
</text>
```
The code action will put the DOCTYPE declaration after the first `<text>` tag, on the second line:

make-DTD-entity-code-action

If it makes sense to fix this issue, then we should test for it here.

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/EntityNotDeclaredCodeAction.java [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+		if (document.hasProlog()) {
+			insertPosition = new Position(1, 0);
+		} else {
+			insertPosition = new Position(0, 0);
+		}
+
+		String delimiter = document.lineDelimiter(insertPosition.getLine());
+		String message = "Declare DOCTYPE containing ENTITY " + entityName;
+		StringBuilder insertString = new StringBuilder();
+		insertString.append("<!DOCTYPE article [" + delimiter);
```
[the remark this answers is not in the document]
Comment (xorye) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/EntityNotDeclaredCodeAction.java]: Thanks, fixed. I think its a much better idea to get it from the top level element like you said

File: org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/extensions/contentmodel/DTDDiagnosticsTest.java (hunk 3 above)
Comment (xorye) [reply org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/extensions/contentmodel/DTDDiagnosticsTest.java:440]: Thanks, fixed

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/EntityNotDeclaredCodeAction.java (hunk 1 above)
Comment (angelozerr) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/EntityNotDeclaredCodeAction.java:177]: Indeed it's little ugly but for the moment we have not teh choice to do that. Please add a comments for that and set the link to the LSP issue (and please do that for the code #681)

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/EntityNotDeclaredCodeAction.java (hunk 2 above)
Comment (angelozerr) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/EntityNotDeclaredCodeAction.java]: To generate XML, it should be better to use XMLBuilder instead of StringBuilder.

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/dom/DOMDocument.java [hunk 5]
```diff
@@ -307,6 +307,24 @@ public boolean hasProlog() {
...
 	}
 
+	/**
+	 * Returns true if this document has a DTD DOCTYPE
+	 * declaration. Returns false otherwise.
+	 * 
+	 * @return true if this document has a DTD DOCTYPE
+	 * declaration. Returns false otherwise.
+	 */
+	public boolean hasDoctype() {
```
Comment (angelozerr) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/dom/DOMDocument.java:317]: I think you don"t need this method. Please remove it and use DOMDocumentType docType = document.getDoctype();.

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/EntityNotDeclaredCodeAction.java [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+	public void doCodeAction(Diagnostic diagnostic, Range range, DOMDocument document, List<CodeAction> codeActions,
+			SharedSettings sharedSettings, IComponentProvider componentProvider) {
+
+		try {
+			String entityName = getEntityName(diagnostic, range, document);
+
+			if (entityName == null) {
+				return;
+			}
+			if (document.hasDoctype()) {
```
Comment (angelozerr) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/EntityNotDeclaredCodeAction.java:51]:
```
DOMDocumentType docType = document.getDoctype();
if (docType != null) {
```

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/dom/DOMDocumentType.java [hunk 7]
```diff
@@ -212,7 +211,7 @@ void setSystemId(int start, int end) {
 	 * getStart() to make them relative to 'content'
 	 */
 	public String getSubstring(int start, int end) {
-		return getContent().substring(start - getStart(), end - getStart());
+		return getTextContent().substring(start - getStart(), end - getStart());
```
Comment (angelozerr) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/dom/DOMDocumentType.java:214]: It seems that getTextContent  could return null.

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/EntityNotDeclaredCodeAction.java [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+
+				// Case 1: <!DOCTYPE exists
+				// Add <!ENTITY nbsp "entity-value"> in the subset.
+				// If the subset does not exist, add subset too
+				
+				// ie:
```
Comment (angelozerr) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/EntityNotDeclaredCodeAction.java:57]: I think you have already spoken about the following usecase, but what about DOCTYPE without subset?

**Comment by fbricon** [comment]
please rebase against master

**Comment by xorye** [comment]
Thank you, fixed

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/EntityNotDeclaredCodeAction.java [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+			} else {
+				// Case 2: <!DOCTYPE does not exist
+				// Generate:
+				// <!DOCTYPE article [
+				//     <!ENTITY nbsp "entity-value">
+				// ]>
+				addDoctypeAndEntityCodeAction(entityName, diagnostic, document, sharedSettings, codeActions);
+			}
+		} catch (BadLocationException e) {
+
```
Comment (angelozerr) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/EntityNotDeclaredCodeAction.java:73]: Add LOGGER

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/EntityNotDeclaredCodeAction.java [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+		}
+	}
+
+	private void addEntityCodeAction(String entityName, Diagnostic diagnostic, DOMDocument document,
```
Comment (angelozerr) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/EntityNotDeclaredCodeAction.java:77]: add Javadoc

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/EntityNotDeclaredCodeAction.java [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+
+		DOMDocumentType docType = document.getDoctype();
+		Position docTypeEnd = document.positionAt(docType.getEnd());
+		Position insertPosition = getEntityInsertPosition(document);
+		String message = "Declare ENTITY " + entityName;
+		String delimiter = document.lineDelimiter(insertPosition.getLine());
+
+		XMLBuilder insertString = new XMLBuilder(settings, null, delimiter);
+
+		if (docType.getInternalSubset() == null) {
```
Comment (angelozerr) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/EntityNotDeclaredCodeAction.java:88]:
```java
boolean hasInternalSubset = docType.getInternalSubset() != null;
if (!hasInternalSubset ) {
```

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/EntityNotDeclaredCodeAction.java [hunk 12]
```diff
@@ -0,0 +1,211 @@
...
+			}
+		} else if (insertPosition.getCharacter() != 0) {
+			// subset exists
+			insertString.linefeed(); // add new entity in a new line
+		}
+
+		insertString.indent(1);
+		addEntityDeclaration(entityName, insertString);
+
+		if (docType.getInternalSubset() == null) {
```
Comment (angelozerr) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/EntityNotDeclaredCodeAction.java:109]: `if (!hasInternalSubset ) {`

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/EntityNotDeclaredCodeAction.java [hunk 13]
```diff
@@ -0,0 +1,211 @@
...
+		} else if (docTypeEnd.getLine() == insertPosition.getLine()) {
+			insertString.linefeed();
+		}
+
+		CodeAction action = CodeActionFactory.insert(message, insertPosition, insertString.toString(),
+				document.getTextDocument(), diagnostic);
+		codeActions.add(action);
+	}
+
+	private void addDoctypeAndEntityCodeAction(String entityName, Diagnostic diagnostic, DOMDocument document,
```
Comment (angelozerr) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/EntityNotDeclaredCodeAction.java:120]: Add Javadoc

**Comment by xorye** [comment]
Fixed, I also added a second commit that adds tests for https://github.com/eclipse/lemminx/issues/686, since that issue is being fixed thanks to this new method:
https://github.com/eclipse/lemminx/blob/765bc8fe1959f2f54bea9667e6c4b8e6320ea450/org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/utils/XMLBuilder.java#L423-L432

File: org.eclipse.lemminx/src/main/resources/org/eclipse/lemminx/services/snippets/pi-snippets.json [hunk 14]
```diff
@@ -0,0 +1,20 @@
+{
```
Comment (angelozerr) [inline org.eclipse.lemminx/src/main/resources/org/eclipse/lemminx/services/snippets/pi-snippets.json:1]: I think this file should not belong to this PR

Comment (xorye) [reply org.eclipse.lemminx/src/main/resources/org/eclipse/lemminx/services/snippets/pi-snippets.json]: thanks, removed

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/utils/StringUtils.java [hunk 15]
```diff
@@ -439,4 +439,47 @@ public static int findEndWord(String text, int offset, Predicate<Character> isVa
...
+			return false;
+		}
+		if (value.isEmpty()) {
+			return false;
+		}
+		char quoteValueStart = value.charAt(0);
+		boolean start = quoteValueStart == '\"' || quoteValueStart == '\'' ? true : false;
+		if (start == false) {
+			return false;
+		}
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/utils/StringUtils.java:459]: +1

**Comment by xorye** [comment]
Great feedback, fixed

**Comment by angelozerr** [comment]
@xorye you have 2 commits for this PR. I think the commit https://github.com/eclipse/lemminx/pull/724/commits/084c611c35645c459a144980c03acaec2af013f3 should be removed, no?

**Comment by fbricon** [comment]
@angelozerr see https://github.com/eclipse/lemminx/pull/724#issuecomment-636011370

**Comment by angelozerr** [comment]
> @angelozerr see #724 (comment)

Thanks @fbricon

**Comment by angelozerr** [comment]
Great PR @xorye !
